docs2/site/docs/getting-started/custom-scalars.md

```diff
@@ -183,105 +355,310 @@ public class ExampleSchema : Schema
+        _ => ThrowValueConversionError(value)
+    }
 
+    public virtual bool CanParseValue(object value)
+    {
+        try
+        {
+            _ = ParseValue(value);
+            return true;
+        }
+        catch
+        {
+            return false;
+        }
+    }
+
+    public virtual object Serialize(object value) => ParseValue(value);
+
+    public override IValue ToAST(object value) => Serialize(value) switch
+    {
+        bool b => new BooleanValue(b),
+        null => new NullValue(),
+        _ => ThrowASTConversionError(value)
+    };
+}
 ```
 
-This is necessary since the query executor converts all arguments and variables to `IValue` instances before coercing them to their server-side representation using `ParseLiteral`. `Vector3` instances parsed from variables will be converted to the more specific `Vector3Value` type.
+### 3. Register the custom scalar within your schema.
 
-7. Implement `ScalarGraphType.Serialize` so `Vector3` instances can be sent to the client.
+The final step is to register an instance of the custom scalar to the schema. This can be
+done for code-first or schema-first schemas. For code-first schemas, register it within
+your constructor via `RegisterType`, as follows:
 
 ```csharp
-// In Vector3Type
-
-public override object Serialize(object value)
+public class MySchema : Schema
 {
-    return ValueConverter.ConvertTo(value, typeof(Vector3));
+    public void MySchema()
+    {
+        Query = ....;
+
+        RegisterType(new MyBooleanGraphType());
+    }
 }
 ```
 
-This implementation may surprise you. Why is `Serialize`, which is used for output, implemented identically to `ParseValue`, which is used for input? Why does `Serialize` return an object, rather than a string or byte array? It helps to understand a few internals of the library.
+For schema-first schemas, register it immediately after calling `Schema.For` to create the schema.
 
-- `Serialize` will be called during query execution, and should be passed an instance of `Vector3` from a field resolver.
+```csharp
+var schema = Schema.For(...);
+schema.RegisterType(new MyBooleanGraphType());
+```
 
-- `Serialize` is _also_ called when reading variables from the client so that variables can be converted to `IValue` instances. In the case of `Vector3Type`, `value` will be a string during this process.
+Now all `BooleanGraphType` references in your schema will utilize the new `MyBooleanGraphType`
+registered with the schema. This technique can be used to replace any of the built-in graph types.
 
-- `ValueConverter.ConvertTo` handles the case when `value` is an instance of the requested type by returning `value`. Therefore, `ValueConverter.ConvertTo` neatly handles both input and output representations of the scalar.
+Note that if you set the `ResolvedType` property of a field or argument to an instance of a built-in
+type, or provide an instance of a built-in type to an applicable constructor, it will not be replaced
```

Comment:
> it will **not be** replaced with your registered replacement built-in type
> This is **by design**.

😮 It was a surprise for me. Why is that?

Reply:
Setting the `ResolvedType` is, by design, supposed to use the exact instance of the graph type that was passed to it.  This is necessary when writing code like this:

```csharp
var s = new Schema();
var query = new ObjectGraphType();
query.Name = "Query";
query.Field(...);
s.Query = query;
var mutation = new ObjectGraphType();
mutation.Name = "Mutation";
...
s.Mutation = mutation;
```

Now scalars typically don't have any properties on them, so they don't typically change as in the above example.

But the `RegisterType` is designed to, basically, register a type that can be used when setting all the `ResolvedType` instances throughout the schema.  If an object already has the `ResolvedType` property set, it should not be reset.

So that's why I say that it is by design.  Due to this design, we could change all the introspection types to be 'locked' onto the built-in scalar types by simply setting `ResolvedType` during their initialization, ensuring they always work the same regardless of the user's schema.  The same could be done by users within their own schema if there was a necessary place to do so.

If someone needs to change ALL scalars of a certain type after the schema is built, they can use `ReplaceScalar` afterwards as documented below.

Reply:
I think I understand. So you are talking about `if ([field/arg].ResolvedType != null) { set_resolved_type }` blocks inside `SchemaTypes.cs`. Right?

Reply:
Thus, the `ReplaceScalar` method takes its place because it is necessary :)